File: tools/commit-benchmark/mandelbrot.cpp
#include "sel.hpp"
#include <chrono>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <iterator>
using Clock=std::chrono::steady_clock;
int main(int argc,char**argv){
 if(argc!=2)return 2;
 std::ifstream input("examples/mandelbrot.sel");std::string source((std::istreambuf_iterator<char>(input)),{});
 auto t=Clock::now();auto p=sel::compile(source);double compile_ms=std::chrono::duration<double,std::milli>(Clock::now()-t).count();
 std::ofstream out(argv[1]);out<<"{\"compile_ms\":"<<compile_ms<<",\"samples_ms\":[";
 int warmups=std::getenv("MANDEL_WARMUPS")?std::atoi(std::getenv("MANDEL_WARMUPS")):2;
 int runs=std::getenv("MANDEL_RUNS")?std::atoi(std::getenv("MANDEL_RUNS")):5;
 std::string outputs;
 for(int i=0;i<warmups+runs;i++){t=Clock::now();auto result=p.run();auto output=result.as_text();double ms=std::chrono::duration<double,std::milli>(Clock::now()-t).count();if(i>=warmups){if(i>warmups){out<<',';outputs+=',';}out<<ms;outputs+=sel::Value::text(output).dump().substr(1);}}
 out<<"],\"outputs\":["<<outputs<<"],\"warmups\":"<<warmups<<",\"runs\":"<<runs<<"}";
}

File: tools/commit-benchmark/mandelbrot.mjs
import {readFileSync,writeFileSync} from 'node:fs';
import {pathToFileURL} from 'node:url';
const {compile,Value}=await import(pathToFileURL(process.cwd()+'/js/src/sel.mjs'));
const source=readFileSync('examples/mandelbrot.sel','utf8');
let t=performance.now(); const program=compile(source); const compile_ms=performance.now()-t;
const samples_ms=[],outputs=[];
const warmups=Number(process.env.MANDEL_WARMUPS ?? 2),runs=Number(process.env.MANDEL_RUNS ?? 5);
for(let i=0;i<warmups+runs;i++) { t=performance.now(); const result=program.run(Value.none()); const output=result.asText(); const elapsed=performance.now()-t; if(i>=warmups){samples_ms.push(elapsed);outputs.push(output);} }
writeFileSync(process.argv[2],JSON.stringify({compile_ms,samples_ms,outputs,warmups,runs}));
